internal/chain/supply.go

```diff
@@ -0,0 +1,38 @@
+package chain
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/harmony-one/harmony/consensus/engine"
+	"github.com/harmony-one/harmony/consensus/reward"
+	"github.com/harmony-one/harmony/numeric"
+	"github.com/harmony-one/harmony/shard"
+	stakingReward "github.com/harmony-one/harmony/staking/reward"
+)
+
+// GetCirculatingSupply using the following formula:
+// (TotalPreStakingTokens * percentReleased) + StakingBlockRewards
+//
+// WARNING: only works on beaconchain if in staking era.
+func GetCirculatingSupply(
+	ctx context.Context, chain engine.ChainReader,
+) (ret numeric.Dec, err error) {
+	currHeader, timestamp := chain.CurrentHeader(), time.Now().Unix()
+	stakingBlockRewards := big.NewInt(0)
+
+	if chain.Config().IsStaking(currHeader.Epoch()) {
+		if chain.ShardID() != shard.BeaconChainShardID {
+			return numeric.Dec{}, stakingReward.ErrInvalidBeaconChain
+		}
+		if stakingBlockRewards, err = chain.ReadBlockRewardAccumulator(currHeader.Number().Uint64()); err != nil {
+			return numeric.Dec{}, err
+		}
+	}
+
+	releasedPreStakingSupply := stakingReward.TotalPreStakingTokens.Mul(
```

Comment:
Actually the unlock schedule is based on the 12.6B token, as you can see from https://docs.google.com/spreadsheets/d/143qpsVaezh9sY_pDhNK-cffwgEdRojWa38-rqBwKI7c/edit#gid=1137842585. The unlock schedule does not concern the block rewards in pre staking era. So `reward.PercentageForTimeStamp(timestamp)` should simply apply on 12.6B and then the pre-staking block reward can be added on top.

Reply:
Changed, some structures to reflect this.